### ISSUE TITLE
performance gain when kernel operation is wrapped with a function

### ISSUE THREAD

**Issue by johnnychen94** [JuliaLang/julia#37558]
This originates from https://github.com/JuliaArrays/OffsetArrays.jl/issues/136#issuecomment-691699933

```julia
function overflow_check(r, offset::T) where T
    if offset > 0 && last(r) > typemax(T) - offset
        throw(ArgumentError("Boundary overflow detected: offset $offset should be equal or less than $(typemax(T) - last(r))"))
    end
end

function overflow_check_with_func(r, offset::T) where T
    throw_overflow_error() = throw(ArgumentError("Boundary overflow detected: offset $offset should be equal or less than $(typemax(T) - last(r))"))
    if offset > 0 && last(r) > typemax(T) - offset
        throw_overflow_error()
    end
end
```

These two works equivalently, but their performance is slightly different:

```julia
julia> r, offset = -3:3, -4;

julia> @btime overflow_check($r, $offset)
  4.801 ns (0 allocations: 0 bytes) # 1.6.0-DEV.883
  4.078 ns (0 allocations: 0 bytes) # 1.4.0, 1.5.1

julia> @btime overflow_check_with_func($r, $offset)
  3.765 ns (0 allocations: 0 bytes) # 1.6.0-DEV.883
  1.696 ns (0 allocations: 0 bytes) # 1.5.1
  1.919 ns (0 allocations: 0 bytes) # 1.4.0
```

I'm really unsure what an appropriate title should be, please rename it if that's unclear.

cc: @jishnub @stevengj

**Comment by timholy** [comment]
Related: #33273, perhaps #22423. 1.6 allocates a gcframe even if you put a `@noinline` in front of `throw_overflow_error`.

**Comment by JeffBezanson** [comment]
Probably due to #35982.

**Comment by JeffBezanson** [comment]
I think we might be able to fix this by removing just the part that treats unreachable nodes the same as throw calls, so that we can still optimize the call to `throw_overflow_error`.

**Comment by JeffBezanson** [comment]
To be clear, that would fix the regression since 1.5, but it's not easy to fix the need to manually out-line the error function.

**Comment by yuyichao** [comment]
> To be clear, that would fix the regression since 1.5, but it's not easy to fix the need to manually out-line the error function.

If it's only about the gc frame that's in the pipeline.
